File: algorithms/dynamic-programming/digit-dp/cpp/digit_dp.cpp
#include <cstring>
#include <string>

namespace {
int digits[12];
int digit_count;
int target_sum;
int memo[12][110][2];

int solve(int position, int current_sum, int tight) {
    if (current_sum > target_sum) {
        return 0;
    }
    if (position == digit_count) {
        return current_sum == target_sum ? 1 : 0;
    }
    if (memo[position][current_sum][tight] != -1) {
        return memo[position][current_sum][tight];
    }

    int limit = tight ? digits[position] : 9;
    int count = 0;
    for (int digit = 0; digit <= limit; ++digit) {
        count += solve(position + 1, current_sum + digit, tight && digit == limit);
    }
    memo[position][current_sum][tight] = count;
    return count;
}
}  // namespace

int digit_dp(int n, int target) {
    if (n <= 0 || target < 0) {
        return 0;
    }

    target_sum = target;
    std::string value = std::to_string(n);
    digit_count = static_cast<int>(value.size());
    for (int index = 0; index < digit_count; ++index) {
        digits[index] = value[index] - '0';
    }

    std::memset(memo, -1, sizeof(memo));
    int count = solve(0, 0, 1);
    if (target == 0) {
        count -= 1;  // Exclude zero because the tests count from 1..N.
    }
    return count;
}

File: algorithms/dynamic-programming/digit-dp/go/digit_dp.go
package main

import (
	"fmt"
	"strconv"
)

var (
	digits    []int
	numDigits int
	targetSum int
	memo      [12][110][2]int
)

func solve(pos, currentSum, tight int) int {
	if currentSum > targetSum {
		return 0
	}
	if pos == numDigits {
		if currentSum == targetSum {
			return 1
		}
		return 0
	}
	if memo[pos][currentSum][tight] != -1 {
		return memo[pos][currentSum][tight]
	}

	limit := 9
	if tight == 1 {
		limit = digits[pos]
	}
	result := 0
	for d := 0; d <= limit; d++ {
		newTight := 0
		if tight == 1 && d == limit {
			newTight = 1
		}
		result += solve(pos+1, currentSum+d, newTight)
	}

	memo[pos][currentSum][tight] = result
	return result
}

func digitDp(n, target int) int {
	if n <= 0 {
		return 0
	}
	targetSum = target

	s := strconv.Itoa(n)
	numDigits = len(s)
	digits = make([]int, numDigits)
	for i := 0; i < numDigits; i++ {
		digits[i] = int(s[i] - '0')
	}

	for i := range memo {
		for j := range memo[i] {
			for k := range memo[i][j] {
				memo[i][j][k] = -1
			}
		}
	}

	result := solve(0, 0, 1)
	if target == 0 {
		// Exclude zero because the contract counts numbers in 1..N.
		result--
	}
	return result
}

func main() {
	var n, target int
	fmt.Scan(&n, &target)
	fmt.Println(digitDp(n, target))
}
